### PR TITLE
Cisco asa

### PR DESCRIPTION
**Pull request by orlichter1**
##4184  Status
In Hold - Instance crashed. Waiting for fix - Details in issue. 

## Related Issues
fixes: https://github.com/demisto/etc/issues/5495

## Description
A new integration for Cisco ASA.
## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](https://github.com/demisto/content/edit/Cisco-ASA/Integrations/CiscoASA/CiscoASA.yml?pr=%2Fdemisto%2Fcontent%2Fpull%2F4585)
- [ ] [CHANGELOG](link)
- [ ] [Description] (https://github.com/demisto/content/edit/Cisco-ASA/Integrations/CiscoASA/CiscoASA_description.md?pr=%2Fdemisto%2Fcontent%2Fpull%2F4585)
- [ ] [Readme] (https://github.com/demisto/content/edit/Cisco-ASA/Integrations/CiscoASA/README.md?pr=%2Fdemisto%2Fcontent%2Fpull%2F4585)

not to forget to merge: https://github.com/demisto/content-test-conf/pull/647

### CODE REVIEW

File: Integrations/CiscoASA/CiscoASA.py [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+                                      "value": source}
+    elif source == 'any':
+        rule_body["sourceAddress"] = {"kind": "AnyIPAddress",
+                                      "value": "any4"}
+    elif '/' in source:
+        rule_body["sourceAddress"] = {"kind": "IPv4Network",
+                                      "value": source}
+    else:
+        rule_body["sourceAddress"] = {"kind": "objectRef#NetworkObj",
+                                      "objectId": source}
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]: move to a function, you have this code 4 times

File: Integrations/CiscoASA/CiscoASA.yml [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+- defaultvalue: 'false'
+  display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- display: is ASAv
+  name: isASAv
+  required: false
+  type: 8
+description: Manage Cisco ASA
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.yml]: elaborate more about the capabilities here (also what ASA stands for)

File: Integrations/CiscoASA/CiscoASA.yml [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+  display: Use system proxy settings
+  name: proxy
+  required: false
+  type: 8
+- defaultvalue: 'false'
+  display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- display: is ASAv
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.yml]: add tool tip for this one
under `additionalinfo` key

**Comment by orlichter1** [comment]
@yaakovi fixed everything except the description of integration which I asked Arseny to formulate.

File: Integrations/CiscoASA/CiscoASA.py [hunk 4]
```diff
@@ -539,12 +513,8 @@ def main():
 
     # Log exceptions
     except Exception as e:
-        LOG.print_log()
         return_error(f"Failed to execute {demisto.command()} command. Error: {e}")
         raise
-    finally:
-        if isASAv and client:
-            client._http_request('DELETE', f'/api/tokenservices/{auth_token}', resp_type='response')
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]: `client.logout()`

File: Integrations/CiscoASA/CiscoASA.py [hunk 5]
```diff
@@ -17,6 +16,20 @@
...
+
+    def login(self, isASAv) -> None:
+        if isASAv:
+            self.isASAv = True
+            res = self._http_request('POST', '/api/tokenservices', resp_type='response')
+            auth_token = res.headers._store.get('x-auth-token')[1]
+            self._headers['X-Auth-Token'] = auth_token
+            self.auth_token = auth_token
+
+    def __del__(self):
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]:
```suggestion
    def logout(self):
```

File: Integrations/CiscoASA/CiscoASA.py [hunk 6]
```diff
@@ -539,12 +513,8 @@ def main():
 
     # Log exceptions
     except Exception as e:
-        LOG.print_log()
         return_error(f"Failed to execute {demisto.command()} command. Error: {e}")
         raise
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]: remove

**Comment by yaakovi** [comment]
@Arsenikr any update on the description for the integration

File: Integrations/CiscoASA/CiscoASA.yml (hunk 2 above)
Comment (Arsenikr) [reply Integrations/CiscoASA/CiscoASA.yml]: Cisco Adaptive Security Appliance Software is the core operating system for the Cisco ASA Family. It delivers enterprise-class firewall capabilities for ASA devices

**Comment by yaakovi** [comment]
@orlichter1 
see arseny cooment regarding description: https://github.com/demisto/content/pull/4585#discussion_r390864770

File: Integrations/CiscoASA/CiscoASA.py (hunk 5 above)
Comment (orlichter1) [reply Integrations/CiscoASA/CiscoASA.py]: why not use __del__?

File: Integrations/CiscoASA/CiscoASA.py [hunk 7]
```diff
@@ -462,6 +496,48 @@ def edit_rule_command(client: Client, args):
...
+def list_interfaces_command(client: Client, args: dict):
+    raw_interfaces = client.list_interfaces()
+    interface_list = []
+    for interface in raw_interfaces:
+
+        temp_interface = {'Type': interface.get('direction', '').capitalize(),
+                          'ID': interface.get('interface', {}).get('objectId', '-1'),
+                          'Name': interface.get('interface', {}).get('name')}
+        interface_list.append(temp_interface)
+    ec = {'CiscoASA.Interface(val.ID == obj.ID)': interface_list}
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]:
```suggestion
    ec = {'CiscoASA.Interface(val.ID && val.ID == obj.ID)': interface_list}
```

File: Integrations/CiscoASA/CiscoASA.py [hunk 8]
```diff
@@ -462,6 +496,48 @@ def edit_rule_command(client: Client, args):
...
+    hr = tableToMarkdown('Interfaces', interface_list, ['Type', 'ID', 'Name'])
+    return hr, ec, hr
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]:
```suggestion
    return hr, ec, raw_interfaces
```

File: Integrations/CiscoASA/CiscoASA.py [hunk 9]
```diff
@@ -462,6 +496,48 @@ def edit_rule_command(client: Client, args):
...
+    for object in objects:
+        if (not obj_names and not obj_ids) or object.get('name') in obj_names or object.get('objectId') in obj_ids:
+            object.pop('selfLink')
+            object.pop('kind')
+            formated_obj = camelize(object)
+            formated_obj['ID'] = formated_obj.pop('Objectid')
+            formated_objects.append(formated_obj)
+    ec = {'CiscoASA.NetworkObject(val.ID == obj.ID)': formated_objects}
+    hr = tableToMarkdown("Network Objects", formated_objects, headers=['ID', 'Name', 'Host', 'Description'])
+    return hr, ec, json.dumps(formated_objects)
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]:
```suggestion
    return hr, ec, formated_objects
```

File: Integrations/CiscoASA/CiscoASA.py [hunk 10]
```diff
@@ -462,6 +496,48 @@ def edit_rule_command(client: Client, args):
         else:
             raise e
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]:
```suggestion
            raise
```

File: Integrations/CiscoASA/CiscoASA.py [hunk 11]
```diff
@@ -157,6 +162,35 @@ def backup(self, data: dict):
...
+                  "host": {
+                    "kind": OBJECT_TYPES_DICT.get(obj_type),
+                    "value": obj_value
+                  }
+                }
+        try:
+            return self._http_request('POST', '/api/objects/networkobjects', json_data=data, ok_codes=(200,201,204),
+                                  resp_type='response')
+        except Exception:
+            raise ValueError("This network object already exists.")
```
Comment (yaakovi) [inline Integrations/CiscoASA/CiscoASA.py]: this is the only possible failure? seems wrong

**Comment by orlichter1** [comment]
@kirbles19 Could you please go over this? Thanks :)

**Comment by kirbles19** [comment]
@orlichter1 Done.

**Comment by jochman** [comment]
@orlichter1 what is the status here?

**Comment by orlichter1** [comment]
@jochman Changed the README according to Alex's changes. Once circle deems it ok I'll merge it

**Pull request closed without merging**